Update esp-idf to v4.3

askpatrickw opened the pull request:
Starting the work for #4182 

- Updated the branch ESP-IDF to point to releases/v4.3
- Fixed the include paths in Makefile 
- Fixed #includes which had incorrect full paths to use shorter paths

Open  Questions:
- There are other full path includes.. they work, but should they be changed?
- Local build now fails on actual code changes which would be needed.

This is the first one, but there are more errors behind this. It would be great if someone who knows C could pickup from here.

```
common-hal/busio/I2C.c: In function 'common_hal_busio_i2c_construct':
common-hal/busio/I2C.c:93:56: error: comparison of distinct pointer types lacks a cast [-Werror]
     if (xSemaphoreCreateBinaryStatic(&self->semaphore) != &self->semaphore) {
                                                        ^~
In file included from ./common-hal/busio/I2C.h:34,
                 from ../../shared-bindings/busio/I2C.h:40,
                 from common-hal/busio/I2C.c:27:
common-hal/busio/I2C.c: In function 'common_hal_busio_i2c_try_lock':
common-hal/busio/I2C.c:169:37: error: passing argument 1 of 'xQueueSemaphoreTake' from incompatible pointer type [-Werror=incompatible-pointer-types]
     self->has_lock = xSemaphoreTake(&self->semaphore, 0) == pdTRUE;
                                     ^
In file included from esp-idf/components/freertos/include/freertos/semphr.h:35,
                 from ./common-hal/busio/I2C.h:34,
                 from ../../shared-bindings/busio/I2C.h:40,
                 from common-hal/busio/I2C.c:27:
esp-idf/components/freertos/include/freertos/queue.h:1239:47: note: expected 'QueueHandle_t' {aka 'struct QueueDefinition *'} but argument is of type 'StaticSemaphore_t *' {aka 'struct xSTATIC_QUEUE *'}
 BaseType_t xQueueSemaphoreTake( QueueHandle_t xQueue, TickType_t xTicksToWait ) PRIVILEGED_FUNCTION;
                                 ~~~~~~~~~~~~~~^~~~~~
cc1: all warnings being treated as errors
make: *** [build-unexpectedmaker_feathers2/common-hal/busio/I2C.o] Error 1
```

skieast commented:
I also started this chase down the rabbit hole.  I however did not make nice edits so my result is not really pretty. 
The semaphore stuff is easy to fix. Once that is done the SPI requires some relatively major changes as Espressif have refactored some calls. I made some changes so it compiles but doesnt work.   I will now start again :) 

I'm not really liking all the full path includes. I changed them to most;ly match how the IDF does its includes. I think this might make it easier for future IDF releases. 

I had started this to see if the newer IDF release fixed anything with I2C, havent got that far yet.

askpatrickw commented:
Thanks @skieast, I appreciate  you picking this up and running with it !

skieast commented:
> 
> 
> Thanks @skieast, I appreciate you picking this up and running with it !

tbh I have no idea how to 'run with it'.   Does that somehow involve me modifying your PR? Or do I just start a new PR with what I have.   As you can probably see my github knowledge is limited.  Very much so. 
I may just start a new local branch and make my changes in a more consistent way now that I know it works.

askpatrickw commented:
@skieast  I think you can use the gh cli to checkout the PR.
gh pr checkout 4195  But maybe you need to be a maintainer on my fork ... hmmm no sure now that I say that outloud.

skieast commented:
I'll do some reading. Forgot about the gh cli. In any case I'm going to at least try to get spi working here. Thanks.

microDev1 commented:
I should have made a PR earlier lol... 
I have a compiling version available in my [`update-idf`](https://github.com/adafruit/circuitpython/compare/main...microDev1:update-idf) branch.
Didn't make PR because SPI and I2C stuff needed some testing. How should we proceed now?

skieast commented:
> 
> 
> I should have made a PR earlier lol...
> I have a compiling version available in my [`update-idf`](https://github.com/adafruit/circuitpython/compare/main...microDev1:update-idf) branch.
> Didn't make PR because SPI and I2C stuff needed some testing. How should we proceed now?

Awesome. I will take a look at your branch. If I have time later tomorrow I'll do some i2c testing.

askpatrickw commented:
On the Discord it was suggested you could fork my fork, but I know you can't do that really without removing other forks. I could make you collaborators on  my fork instead.

But, why don't we abandon this PR and you can do your own sounds like @microDev1 had already gotten farther than I had anyway. That's probably the easiest. I'll leave this open and you all can decide. Thanks for doing the work for this!

hierophect commented:
How's the progress on this coming along? I'd like to update in order to try and make more progress on the ongoing I2C problems. To clone someone else's branch, you can follow the instructions [here](https://gist.github.com/elfrank/c08256de9c15e41e1781), it's not too bad, but you do need to make sure your PR is marked to accept contributions.

skieast commented:
I’m out now. As I posted on discord the branch from microdev compiles but had some issues. I can get i2c scan to work but otherwise it is very unstable. Will try some code later to see if crashing is general.

tannewt commented:
> I'm not really liking all the full path includes. I changed them to most;ly match how the IDF does its includes. I think this might make it easier for future IDF releases.

Why not? I like the full path because that makes them easier to find. I dislike having super long include directory lists and include filenames that look like they come from the same place but don't.

> On the Discord it was suggested you could fork my fork, but I know you can't do that really without removing other forks. I could make you collaborators on my fork instead.

You can always add each other as remotes to your local repo:

`git remote add <remote name> <git url>`

askpatrickw commented:
> > I'm not really liking all the full path includes. I changed them to most;ly match how the IDF does its includes. I think this might make it easier for future IDF releases.
> 
> Why not? I like the full path because that makes them easier to find. I dislike having super long include directory lists and include filenames that look like they come from the same place but don't.

I found this very confusing as well. I would suggest the time might be right for having a philosophy (I'm shying away from the word "Rule") about this and strive for greater consistency.

microDev1 commented:
A follow-up with the current status :-
- SPI & I2C might have bugs.
- Currently I am focusing on getting things up and running which were stuck due to `idf@v4.2`.  
- I suggest creating a new branch for the IDF update. That way it would be easier for folks to collaborate on the update work.

skieast commented:
Yes. Good idea. I think my fixes to i2c from your branch have that working. No testing on SPI yet. Like you I expect bugs. But there's the fun.  I'm currently looking (again) at i2c and wifi crashing. Wifi is not trivial. Lol.

askpatrickw commented:
@microDev1  and @skieast if you want to try ... I added you as contributors to my fork.
Scott suggested you could add my fork as a remote `git remote add askpatrickw https://github.com/askpatrickw/circuitpython.git` and then merge into my branch which is connected to this PR. At least I think so...  ;-)

Whatever is easiest though.. I'm not concerned about it coming from my repo or this PR.

skieast commented:
> 
> 
> @microDev1 and @skieast if you want to try ... I added you as contributors to my fork.
> Scott suggested you could add my fork as a remote `git remote add askpatrickw https://github.com/askpatrickw/circuitpython.git` and then merge into my branch which is connected to this PR. At least I think so... ;-)
> 
> Whatever is easiest though.. I'm not concerned about it coming from my repo or this PR.

Always like to learn new things. I will look at merging my i2c changes. Will have to clean up a bit first as I have some additional boards defined for my use.  Maybe during Scott's Deep Dive today. :)

askpatrickw commented:
@skieast  and @microDev1 , I've been away from CP for a few days. How is the IDF update looking? Is there anything I can help with?

skieast commented:
> 
> 
> @skieast and @microDev1 , I've been away from CP for a few days. How is the IDF update looking? Is there anything I can help with?

I too took a little break after running into a wall with the I2C and WIFI issues.  I've started again and will probably ignore the wifi for a bit.

askpatrickw commented:
One thing I didn't quite understand. Are the i2c and WiFi issues in the current IDF version in CP?

skieast commented:
Yes. I'm not sure when the problem first appeared but it's defnitey there with 4.2. I was hoping that 4.3 would fix or at least change the issue.   As I said I'm going to get i2c working then see if spi works. Those were the two areas where compile errors appeared  that required changes.   Then look at wifi.

askpatrickw commented:
If 4.3 doesn't introduce any _new_ problems, does it makes sense to Merge 4.3 and work on issues not related to IDF 4.3 separately?

microDev1 commented:
Thanks! @askpatrickw and @skieast for looking into this. 
We can deal with issues not related 4.3 later. So far I have only found i2c and spi to be a concern.

askpatrickw commented:
@microDev1 Just to make sure I'm following. The i2c and spi issues are the same ones as before or new ones?

If they are not new, maybe you should make a PR from you branch where you had everything building and we can close this.

skieast commented:
There were changes to the IDF with 4.3 that require changes to i2c and spi.  I2C is mostly done and seems to work as before.  There is an issue with soft reboot / or wifi. You get to pick which one works. I am starting to look at SPI now.  So basically should be the same as before once everything works. SPI had a lot of changes I think so I am hoping whatever microdev did works out of the box.  :)

askpatrickw commented:
Thanks @skieast !

microDev1 commented:
>  I am hoping whatever microdev did works out of the box. :)

It should work, shouldn't it. 🤞
I think only `spi_hal_cal_clock_conf()` has a major change.

skieast commented:
Lol. Crashed when doing the spi construct.  I'm on the ski hill for a few hours and will dig in later. I have an idea what it is. Probably wrong but it's an idea.

askpatrickw commented:
I had a great chat with @unexpectedmaker about this update to the IDF, he's going to post more detailed notes, but a couple of things:

1.The paths updates I provided (now ~ a month old) are incorrect in many instances, wherever possible the path should have an esp32s2 folder in the path.
2. In order to support S3, ESP32 as well as the S2, from the same port, the build system will need to have a way to indicate the target board. @tannewt should advise on that, and we should probably break that out as a separate issue. q: Maybe this can go in the board definition?
3. When the second item is worked on, the conditional paths should also be fixed so either the esp32s2 folder is conditionalyl esp32s3, or esp32 or that there are conditional sections for each target microcontroller's includes

Seon is going to provide a sample of the path changes and if my day allows, I'll give a swing at those changes tomorrow. 

@microDev1 @skieast, I'm also happy to bow out if someone else has time and would like to step-in.

Keep in mind, this also fixes building on the M1 Mac Mini...

askpatrickw commented:
done.

askpatrickw commented:
I updated this pr so it can be the basis for collaboration and moving things forward:

- Moved the esp-idf tag forward to the v4.3-beta1 tag.
- Added IDF_TARGET to every board file, used that in the INC section of the makefile
- Changed a few includes so they use the short path because they had 'esp32s2' in the include.
- Hand merged @microDev1's I2C changes into my branch. There was an I2C build error and then after there was not.
- I alphabetized the INCs. That might be incorrect... let me know.

Known Issues:

1. There are merge conflicts
2. SPI still fails. I tried hand merging Microdev's edits but they didn't resolve the build issue so I did not commit those changes.
3. There is still a lot of esp32s2 specific looking stuff in the  makefile, but this is a start toward mult-mcu IDF builds as discussed in #4363 

As always, let me know if I missed anything or if it could be improved

microDev1 commented:
@askpatrickw I believe these changes should have there own PR after this is merged.
You can simply pull from my branch to incorporate my commit here. That should also fix the merge conflict and build issues.

microDev1 commented:
I have updated the PR as discussed on discord.

askpatrickw reviewed (commented):
I reviewed the changes to the INC in the Makefile and the paths updates in the .c and .h files, they all make sense.
I also built and flashed this successfully to a feathers2 pre-release and ran a program that uses a BME680 over i2c.
Also built and flashed a kaluga and tested an Ultimate GPS over UART which worked as well.

microDev1 commented on .gitmodules:
```diff
@@ -152,7 +152,8 @@
 [submodule "ports/esp32s2/esp-idf"]
 	path = ports/esp32s2/esp-idf
-	url = https://github.com/adafruit/esp-idf.git
+	url = https://github.com/espressif/esp-idf.git
```
I think we will be following the idf releases directly from now on. 
This should avoid confusion and we can switch to a patched version if such a need arises.

skieast replied:
In testing this on my Gravitech Cucumber i get a crash to safe mode with I2C and the SGP30.  I'll try with the UM FeatherS2 to see if this reproduces.

askpatrickw replied:
I was intrigued by your crash, so I ran a very simple test

```python
import board
from busio import I2C
i2c = I2C(board.SCL, board.SDA)
```
FeatherS2 this works fine.
MetroS2, and Kaluga it goes into safemode.

I used the same BME680 on all boards.
Tested my own builds and the artifacts attached to the PR.

The difference between these groups of boards that I see is
the ones that fail  use the WROVER module and the Feather
S2 (mine is a Pre-Release) uses the raw ESP32-S2.

Microdev, if you're testing using your MicroS2, also a raw ESP32-S2, right?

microDev1 replied:
This crashes for me on a debug build but the following fixes it:

```diff
diff --git a/ports/esp32s2/common-hal/busio/I2C.c b/ports/esp32s2/common-hal/busio/I2C.c
index d45d391a2..09e8ce9ca 100644
--- a/ports/esp32s2/common-hal/busio/I2C.c
+++ b/ports/esp32s2/common-hal/busio/I2C.c
@@ -94,7 +94,7 @@ void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
     if (self->semaphore == NULL) {
         mp_raise_RuntimeError(translate("Unable to create lock"));
     }
-    xSemaphoreGive(&self->semaphore);
+    xSemaphoreGive(self->semaphore);
     self->sda_pin = sda;
     self->scl_pin = scl;
     self->i2c_num = I2C_NUM_MAX;
@@ -181,7 +181,7 @@ bool common_hal_busio_i2c_has_lock(busio_i2c_obj_t *self) {
 }
 
 void common_hal_busio_i2c_unlock(busio_i2c_obj_t *self) {
-    xSemaphoreGive(&self->semaphore);
+    xSemaphoreGive(self->semaphore);
     self->has_lock = false;
 }
```

askpatrickw replied:
@microDev1  That works for me on the FeatherS2 and the MetroS2. 🎉

dhalbert replied:
I wonder if this may be the root cause for the I2C/wifi problems. I have a fix for that but it doesn't make a lot of sense in that I am removing code that should be OK.

dhalbert replied:
I am confused by this change. `self->semaphore is a `StaticQueue_t` struct, not a pointer to a such a struct. `xSemaphoreGive()` and `Take()` expect a `QueueHandle_t`, which is a `void *`. `self->semaphore` is initialized by the `xSemaphoreCreateMutexStatic(&self->semaphore)`. So the removal of the `&`s seems incorrect to me.

skieast replied:
I have made some changes to I2C.c and .h to more or less match what my branch has.  Similar to what @dhalbert has suggested.  I'll post diffs once I get back in later today.

dhalbert replied:
That I suggested or that @microDev1 suggested?

askpatrickw replied:
Looks like Skieast made the same changes and Microdev

https://github.com/adafruit/circuitpython/compare/main...skieast:update-idf-v4.3#diff-1dda8b881430b077c387870ff34ca6eebf4a74c16067b736d53aa673dc7d7b8dL93

dhalbert replied:
Could we separate the 4.3 update from the I2C changes? We can keep the buggy I2C as it is for now. I am working on the I2C stuff in https://github.com/adafruit/circuitpython/pull/4387 and https://github.com/adafruit/esp-idf/pull/1.

dhalbert replied:
> Looks like Skieast made the same changes and Microdev
> 
> [main...skieast:update-idf-v4.3diff-1dda8b881430b077c387870ff34ca6eebf4a74c16067b736d53aa673dc7d7b8dL93](https://github.com/adafruit/circuitpython/compare/main...skieast:update-idf-v4.3diff-1dda8b881430b077c387870ff34ca6eebf4a74c16067b736d53aa673dc7d7b8dL93)

I get "there isn't anything to compare" for this link.

askpatrickw replied:
GitHub weirdness... If you navigate to https://github.com/skieast/circuitpython/tree/update-idf-v4.3 and click compare I think it will work.

skieast replied:
Yes, getting very confusing. I will start again with the new I2C.c that appears to work with 4.2. I may have been having some laptop issues that kept blowing up builds.  Rolling back windows to last week seems to have put me back to where I was

microDev1 replied:
I just realized that we are commenting on a review. I am marking this as resolved. Lets continue the conversation below. :)

microDev1 commented:
Converting this to draft for now as `SPI` still crashes. Also, `I2C` will need a re-visit once #4387 is merged.

askpatrickw commented:
#4387  is merged now what should we verify\fix next ?

askpatrickw commented:
ESP-IDF v4.3 released 

https://github.com/espressif/esp-idf/releases/tag/v4.3

microDev1 commented:
Finally... this is almost ready to merge. I re-wrote the `SPI` module, now it utilizes the idf's `spi_master` driver. I haven't tested the new module completely and some bugs might still be present, it would be really helpful if folks can give it a shot. :)

tannewt commented:
Thanks @microDev1 ! The code looks good to me. Unfortunately, it looks like Kaluga 1.3 ran out of space in one language.

Could others chime in that they've tested this too? Thanks!

anecdata commented:
I'm seeing issues with this artifact relative to 7.0.0-alpha.3: ejecting CIRCUITPY and restarting on uncaught exceptions and also on trying to lock I2C:
```
Adafruit CircuitPython 7.0.0-alpha.3-99-gcc7d59c3a on 2021-06-23; FeatherS2 with ESP32S2
>>> import board
>>> i2c = board.I2C()
>>> 
>>> while not i2c.try_lock():
...     pass
...     
...     
... 
$ 
```

microDev1 commented:
I have fixed the Kaluga 1.3 space and `i2c.try_lock()` issues. I haven't noticed any uncaught exceptions yet.

anecdata commented:
Confirmed `i2c.try_lock()` is good now.

Try this code to see the exception crash:
```py
import board
import time

time.sleep(5)

import adafruit_dotstar
# status_light = adafruit_dotstar.DotStar(board.APA102_SCK, board.APA102_MOSI, 1, brightness=0.1)

while True:
    print("", time.monotonic(), "\r", end="")
    time.sleep(1)
```
control-c within the loop works fine unless the DotStar line is uncommented

microDev1 commented:
@anecdata Thanks for the thorough testing, the exception crash is now resolved.

tannewt reviewed (approved):
Thank you for updating the IDF!
